### PR TITLE
Fix main module loading

### PR DESCRIPTION
**Pull request by nunofgs**
### What does this PR do?
Fixes module loading when the package defines a folder as their `main` entrypoint. For, example [pg](https://github.com/brianc/node-postgres/blob/4cb73ebc2c04cd039881a015d623436f26058608/packages/pg/package.json#L20).

### Motivation
I tried to use dd-trace-js to send `pg` traces, but found it didn't work unless I used `pg-native`. After digging into it, I found that the reason were [these lines](https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/platform/node/loader.js#L60-L61) not accounting for packages that don't define a path to a file.

### Additional Notes
- I couldn't find any tests on this functionality but I'm happy to add them if you'd like.
- This only applies if tracing is initiated *after* the module is loaded (ie: loading from the require cache)

### CODE REVIEW

File: packages/dd-trace/src/platform/node/loader.js [hunk 1]
```diff
@@ -57,7 +57,13 @@ class Loader {
 
         pkg = require(`${basedir}/package.json`)
 
-        const mainFile = path.posix.normalize(pkg.main || 'index.js')
+        let mainFile = path.posix.normalize(pkg.main || 'index.js')
+
+        // Support libs that specify a folder as their `main` module, e.g.: `{ "main": "./lib") }`.
+        if (!path.posix.extname(mainFile)) {
```
Comment (rochdev) [inline packages/dd-trace/src/platform/node/loader.js:63]: Node also supports files without extension, so I think the logic should be similar to that of `require` itself.

**Review by rochdev — CHANGES_REQUESTED** [review]
For the tests, this is tested indirectly in the instrumenter tests. Both the `express-mock` and `mysql-mock` modules have a main defined while `other` doesn't. Changing one of them to use a folder would test this.

File: packages/dd-trace/src/platform/node/loader.js (hunk 1 above)
Comment (nunofgs) [reply packages/dd-trace/src/platform/node/loader.js]: Pushed again. Went with `require.resolve()` so as to not reimplement the require logic (which is actually a lot more complex than I had imagined 😅 )

**Comment by nunofgs** [comment]
> For the tests, this is tested indirectly in the instrumenter tests. Both the express-mock and mysql-mock modules have a main defined while other doesn't. Changing one of them to use a folder would test this.

@rochdev thanks for looking into the PR. I reused the `other` test.

File: packages/dd-trace/src/platform/node/loader.js [hunk 2]
```diff
@@ -57,8 +57,18 @@ class Loader {
...
+          const mainFile = path.posix.normalize(pkg.main || 'index.js')
+          const resolvedFile = require.resolve(path.posix.join(basedir, '../', instrumentation.name, mainFile))
+
+          if (!id.endsWith(resolvedFile)) continue
+        } catch (e) {
+          if (e.code === 'MODULE_NOT_FOUND') {
+            continue
+          }
+
+          throw e
```
Comment (rochdev) [inline packages/dd-trace/src/platform/node/loader.js:70]: The tracer should never throw, can the error be logged instead?

File: packages/dd-trace/src/platform/node/loader.js [hunk 3]
```diff
@@ -57,8 +57,18 @@ class Loader {
 
         pkg = require(`${basedir}/package.json`)
 
-        const mainFile = path.posix.normalize(pkg.main || 'index.js')
-        if (!id.endsWith(`/node_modules/${instrumentation.name}/${mainFile}`)) continue
+        try {
+          const mainFile = path.posix.normalize(pkg.main || 'index.js')
+          const resolvedFile = require.resolve(path.posix.join(basedir, '../', instrumentation.name, mainFile))
```
Comment (rochdev) [inline packages/dd-trace/src/platform/node/loader.js:62]: What happens if the module is nested in the `node_modules` of another module imported as a transitive dependency?

**Comment by tlhunter** [comment]
I'm going to close this PR out as it's pretty old and has some unaddressed issues. Plus, I believe the underlying issue might no longer be applicable as we're able to instrument the `pg` module as it exists today.

Feel free to reopen / fix any merge conflicts if you'd like it reviewed again!

**Pull request closed without merging**